src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,143 @@
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size() || index < 0) {
+            throw new IndexOutOfBoundsException(ERROR_MESSAGE_OUT_OF_BOUNDS + index);
+        }
+        if (head == null) {
+            head = new Node<T>(null, value, null);
+            tail = head;
+        } else if (size() == index) {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            newNode.next.prev = newNode;
+            if (index != 0) {
+                newNode.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            tail.next = new Node<>(tail, element, null);
+            tail = tail.next;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T oldValue = nodeByIndex.value;
+        nodeByIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T removedValue = nodeByIndex.value;
+        removeNode(nodeByIndex);
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (Objects.equals(currentNode.value, object)) {
+                removeNode(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        int size = 0;
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            size++;
+            currentNode = currentNode.next;
+        }
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        int currentIndex = 0;
+        while (currentNode != null && index >= 0) {
+            if (index == currentIndex) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
```

Comment:
let's move from tail if index is in the right part of the list

use `size` field to get this information